File: bukkit/teams/captain/src/main/java/dialight/teams/captain/utils/Schematic.java
package dialight.teams.captain.utils;

import dialight.misc.player.UuidPlayer;
import dialight.tuple.Tuple2t;
import org.bukkit.util.Vector;

import java.util.HashMap;
import java.util.Map;

public class Schematic {

    private final Map<Vector, BlockData> blocks = new HashMap<>();
    private final Map<Vector, UuidPlayer> users = new HashMap<>();
    @Deprecated public Vector active;

    public void block(Vector pos, BlockData block) {
        blocks.put(pos, block);
    }

    public void user(Vector pos, UuidPlayer user) {
        users.put(pos, user);
    }

    public Tuple2t<Vector, Vector> getMinMax() {
        int minX = Integer.MAX_VALUE;
        int minY = Integer.MAX_VALUE;
        int minZ = Integer.MAX_VALUE;
        int maxX = Integer.MIN_VALUE;
        int maxY = Integer.MIN_VALUE;
        int maxZ = Integer.MIN_VALUE;
        for (Map.Entry<Vector, BlockData> entry : blocks.entrySet()) {
            Vector vec = entry.getKey();
            BlockData block = entry.getValue();
            if(vec.getX() < minX) minX = vec.getBlockX();
            if(vec.getX() > maxX) maxX = vec.getBlockX();
            if(vec.getY() < minY) minY = vec.getBlockY();
            if(vec.getY() > maxY) maxY = vec.getBlockY();
            if(vec.getZ() < minZ) minZ = vec.getBlockZ();
            if(vec.getZ() > maxZ) maxZ = vec.getBlockZ();
        }
        return new Tuple2t<>(new Vector(minX, minY, minZ), new Vector(maxX, maxY + 2, maxZ));
    }
    public Vector getSize() {
        Tuple2t<Vector, Vector> minMax = getMinMax();
        Vector min = minMax.getT1();
        Vector max = minMax.getT2();
        return max.subtract(min);
    }
    public Vector centerOffset() {
        Tuple2t<Vector, Vector> minMax = getMinMax();
        Vector min = minMax.getT1();
        Vector max = minMax.getT2();
        return min.multiply(-1);
    }

    public Map<Vector, BlockData> getBlocks() {
        return blocks;
    }

    public Map<Vector, UuidPlayer> getUsers() {
        return users;
    }
}

File: teams_controllers/captain/src/main/kotlin/dialight/captain/system/arena/Schematic.kt
package dialight.captain.system.arena

import com.flowpowered.math.vector.Vector3i
import dialight.captain.system.members.CSUser
import org.spongepowered.api.Sponge
import org.spongepowered.api.block.BlockState
import org.spongepowered.api.block.BlockType
import org.spongepowered.api.data.key.Keys
import org.spongepowered.api.data.manipulator.mutable.DyeableData
import org.spongepowered.api.data.type.DyeColor
import java.util.*


class Schematic {
    val blocks = HashMap<Vector3i, BlockState>()
    val users = HashMap<Vector3i, CSUser>()
    lateinit var actve: Vector3i

    fun block(pos: Vector3i, type: BlockType) {
        blocks[pos] = BlockState.builder().blockType(type).build()
    }
    fun block(pos: Vector3i, block: BlockState) {
        blocks[pos] = block
    }
    fun block(pos: Vector3i, type: BlockType, color: DyeColor) {
        val dyeableData = Sponge.getDataManager().getManipulatorBuilder(DyeableData::class.java).get().create()
        dyeableData.set(Keys.DYE_COLOR, color)
        val state = BlockState.builder()
            .blockType(type)
            .add(dyeableData)
            .build()
        blocks[pos] = state
    }

    fun user(vec: Vector3i, user: CSUser) {
        users[vec] = user
    }

    fun getMinMax(): Pair<Vector3i, Vector3i> {
        var minX = Integer.MAX_VALUE
        var minY = Integer.MAX_VALUE
        var minZ = Integer.MAX_VALUE
        var maxX = Integer.MIN_VALUE
        var maxY = Integer.MIN_VALUE
        var maxZ = Integer.MIN_VALUE
        for((vec, block) in blocks.entries) {
            if(vec.x < minX) minX = vec.x
            if(vec.x > maxX) maxX = vec.x
            if(vec.y < minY) minY = vec.y
            if(vec.y > maxY) maxY = vec.y
            if(vec.z < minZ) minZ = vec.z
            if(vec.z > maxZ) maxZ = vec.z
        }
        return Pair(Vector3i(minX, minY, minZ), Vector3i(maxX, maxY, maxZ))
    }
    fun getSize(): Vector3i {
        val (min, max) = getMinMax()
        return max.sub(min)
    }
    fun centerOffset(): Vector3i {
        val (min, max) = getMinMax()
        return min.mul(-1)
    }

}